src/backend/base/langflow/utils/logger.py

```diff
@@ -12,6 +18,95 @@
 VALID_LOG_LEVELS = ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]
 
 
+class SizedLogBuffer:
+    def __init__(self):
+        """
+        a buffer for storing log messages for the log retrieval API
+        the buffer can be overwritten by an env variable LANGFLOW_LOG_RETRIEVER_BUFFER_SIZE
+        because the logger is initialized before the settings_service are loaded
+        """
+        self.max: int = 0
+        env_buffer_size = os.getenv("LANGFLOW_LOG_RETRIEVER_BUFFER_SIZE", "0")
+        if env_buffer_size.isdigit():
+            self.max = int(env_buffer_size)
+        self.buffer: Deque[str] = deque(maxlen=self.max)
+        self._lock = threading.Lock()
+        self.page_size: int = 100
+        self.sessions: Dict[str, Dict] = {}
+        self.session_timeout: timedelta = timedelta(minutes=5)
+
+    def write(self, message: str):
+        record = json.loads(message)
+        log_entry = record["text"]
+        with self._lock:
+            self.buffer.append(log_entry)
+
+    def readlines(self) -> list[str]:
+        return list(self.buffer)
+
+    def enabled(self) -> bool:
+        return self.max > 0
+
+    def max_size(self) -> int:
+        return self.max
+
+    def create_session(self) -> str:
+        session_id = str(uuid.uuid4())
+        self.sessions[session_id] = {
+            "created": datetime.now(),
+            "last_access": datetime.now(),
+            "snapshot": list(self.buffer),
+        }
+        return session_id
+
+    def get_page(self, session_id: str, page: int) -> Dict[str, Any]:
+        self.cleanup_sessions()
+
+        if session_id not in self.sessions:
+            return {"error": "Invalid or expired session ID"}
+
+        session = self.sessions[session_id]
```

Comment:
we need a lock for accessing the sessions

Reply:
removed session